### PR TITLE
Alternative aggregate functions to calculate histogram values.

### PR DESCRIPTION
**Pull request by akalotkin**
*Issue #, if available:*
NA. Kind of close to #381 

*Description of changes:*
- Histogram analyzer changes
  -  Provide a way to calculate histogram values using different aggregate functions (sum).
  - Add aggregateColumn - a column castable to Long. Aggregate function is applied to this column
- Histogram serialization changes
  - Add aggregateFunction property ("count" is default in case of missing property)
  - Add aggregateColumn property

In addition to
``` 
select(col(column).cast(StringType)).groupBy(column).count()
```
provide a way get histogram values
```
select(col(column).cast(StringType), col(aggregateColumn.get).cast(LongType)).groupBy(column).sum(aggregateColumn.get)
```

WIP. Please take a look at the changes. I'll add more tests, docs and validation logic if you think that these changes make sense.

Thanks.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by akalotkin** [comment]
@rdsharma26, can you please take a look? Thanks.

**Comment by rdsharma26** [comment]
@akalotkin Will be having a look shortly. In the meantime, could you fix the build failure?

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 1]
```diff
@@ -43,7 +44,9 @@ case class Histogram(
     binningUdf: Option[UserDefinedFunction] = None,
     maxDetailBins: Integer = Histogram.MaximumAllowedDetailBins,
     where: Option[String] = None,
-    computeFrequenciesAsRatio: Boolean = true)
+    computeFrequenciesAsRatio: Boolean = true,
+    aggregateFunction: String = Histogram.Count,
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:48]: Can we make this an enum? In Scala 2, we can use a `sealed trait` and `case object` to create enumerations.

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 2]
```diff
@@ -43,7 +44,9 @@ case class Histogram(
...
+    aggregateColumn: Option[String] = None)
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:49]: This is used when the aggregate function is sum. As a caller of this API, it is possible that this value is not provided when using `sum`. Is it possible to group this configuration into a single object? 

Check this parameter for reference: https://github.com/awslabs/deequ/commit/ef4c308a165d5928dc61e2349d907a232f8761f9#diff-4db9bbf900d2a35e16c36a158cbf3c4641d336e1930e6f343f2fef5773d0cea6R95

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 3]
```diff
@@ -125,11 +124,33 @@ case class Histogram(
       case _ => data
     }
   }
+
+  private def query(data: DataFrame): DataFrame = {
+    aggregateFunction match {
+      case Histogram.Count => countQuery(data)
+      case Histogram.Sum => sumQuery(data)
+      case _ => throw new IllegalStateException()
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:132]: If we use an enumeration, then we won't need a default case.

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 4]
```diff
@@ -125,11 +124,33 @@ case class Histogram(
...
+    }
+  }
+
+  private def countQuery(data: DataFrame): DataFrame = {
+      data.select(col(column).cast(StringType))
+      .na.fill(Histogram.NullFieldReplacement)
+      .groupBy(column).count().withColumnRenamed("count", Analyzers.COUNT_COL)
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:139]: Can we use the same formatting as the one on lines 66-73 originally?

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 5]
```diff
@@ -125,11 +124,33 @@ case class Histogram(
...
+  }
+
+  private def sumQuery(data: DataFrame): DataFrame = {
+    data.select(col(column).cast(StringType), col(aggregateColumn.get).cast(LongType))
+      .na.fill(Histogram.NullFieldReplacement)
+      .groupBy(column).sum(aggregateColumn.get).withColumnRenamed("count", Analyzers.COUNT_COL)
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:145]: Can we use the same formatting as the one on lines 66-73 originally?

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 6]
```diff
@@ -125,13 +126,115 @@ case class Histogram(
...
+  val sum_function = "sum"
+
+
+  def createAggregateFunction(function: String, aggregateColumn: String): AggregateFunction = {
+    function match {
+      case Histogram.count_function => Count
+      case Histogram.sum_function => Sum(aggregateColumn)
+      case _ => throw new IllegalArgumentException("Wrong aggregate function name: " + function)
+    }
+  }
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:148]: Can we remove this function and let the caller of the function provide a valid instance of `AggregateFunction` ? The reason I say this is because if we add more subclasses of `AggregateFunction` in the future, we will need to update this `createAggregateFunction` to accept the superset of ALL parameters for all subtypes.

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala [hunk 7]
```diff
@@ -125,13 +126,115 @@ case class Histogram(
...
+//        .max(aggColumn)
+//        .withColumnRenamed("count", Analyzers.COUNT_COL)
+//    }
+//
+//    override def aggregateColumn(): Option[String] = {
+//      Some(aggColumn)
+//    }
+//
+//    override def function(): String = max
+//  }
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:232]: Either we remove the commented out code, or we keep it as code and add tests. Otherwise, it adds bloat to the source code.

File: src/main/scala/com/amazon/deequ/repository/AnalysisResultSerde.scala [hunk 8]
```diff
@@ -302,6 +302,10 @@ private[deequ] object AnalyzerSerializer
         result.addProperty(ANALYZER_NAME_FIELD, "Histogram")
         result.addProperty(COLUMN_FIELD, histogram.column)
         result.addProperty("maxDetailBins", histogram.maxDetailBins)
+        if (histogram.aggregateFunction != Histogram.Count) {
```
Comment (rdsharma26) [inline src/main/scala/com/amazon/deequ/repository/AnalysisResultSerde.scala:305]: Can we add a comment here that:
- the reason we are excluding `Count` since it is the current default
- It was the only previously supported function in the past, and we exclude it for backwards compatibility

File: src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala [hunk 9]
```diff
@@ -86,13 +86,13 @@ class AnalyzerContextTest extends AnyWordSpec
               |{"entity":"Column","instance":"item","name":"Distinctness","value":1.0},
               |{"entity":"Column","instance":"att1","name":"Completeness","value":1.0},
               |{"entity":"Multicolumn","instance":"att1,att2","name":"Uniqueness","value":0.25},
+              |{"entity":"Dataset","instance":"*","name":"Size (where: att2 == 'd')","value":1.0},
+              |{"entity":"Dataset","instance":"*","name":"Size","value":4.0},
```
Comment (rdsharma26) [inline src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala:90]: Is this change required? It is not clear from the changes that the order of the fields in the serialization process is changing. If it is not, can we remove this diff to make the PR clearer ?

Comment (akalotkin) [reply src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala:90]: Yes.  This is required. This is the reason a test was failing.

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala (hunk 7 above)
Comment (akalotkin) [reply src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:232]: I guess. I just forgot to remove this code. Will do

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala (hunk 6 above)
Comment (akalotkin) [reply src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:148]: This function is used to deserialize object from json. I can move it to AnalysisResultSerde, but there must be a piece of code that deserializes all combinations from json.

File: src/main/scala/com/amazon/deequ/repository/AnalysisResultSerde.scala (hunk 8 above)
Comment (akalotkin) [reply src/main/scala/com/amazon/deequ/repository/AnalysisResultSerde.scala:305]: Will add comments.

File: src/main/scala/com/amazon/deequ/analyzers/Histogram.scala (hunk 6 above)
Comment (rdsharma26) [reply src/main/scala/com/amazon/deequ/analyzers/Histogram.scala:148]: Sure, I think it is fine then to keep it outside of `Histogram` and in `AnalysisResultSerde`.

File: src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala (hunk 9 above)
Comment (rdsharma26) [reply src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala:90]: Thanks for the information. Is it possible to understand why it is required? The reason I ask is because of backwards compatibility. Will a serialized object that has the older order fail to get deserialized because of this change?

Comment (akalotkin) [reply src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala:90]: I need to dig deeper to explain why it happens. My guess is that there is some kind of HashMap, HashSet that changes order of response.

Comment (akalotkin) [reply src/test/scala/com/amazon/deequ/analyzers/runners/AnalyzerContextTest.scala:90]: Yes. That's what happens. Below are screenshots from 2 different versions (breakpoint is on this line https://github.com/awslabs/deequ/blob/f53283e05ca93a16f054d4783e000082699fe0c3/src/main/scala/com/amazon/deequ/analyzers/runners/AnalysisRunner.scala#L201)

As you can see after map union it gives two different results. And code here (https://github.com/awslabs/deequ/blob/f53283e05ca93a16f054d4783e000082699fe0c3/src/main/scala/com/amazon/deequ/analyzers/runners/AnalyzerContext.scala#L83) iterates through the HashMap. 


Master branch
master
feature/histogram-aggregate branch
feature:histogram-aggregate

**Comment by akalotkin** [comment]
@rdsharma26, all done.

**Comment by rdsharma26** [comment]
Thanks for the PR! @akalotkin

**Review by rdsharma26 — APPROVED** [review]
